### DISCUSSION TITLE
Pattern matching for multidimensional time series

### DISCUSSION THREAD

**Discussion by huberl** [TDAmeritrade/stumpy#630] [Q&A] [answered]
I'm currently working with IMU and other sensor data obtained by a car, which gives we hours of 10-dimensional (i.e. 10 sensors) time series data. I have a target sequence (e.g. driving through a s-curve) and need to find similar occurrences in my data.

I tried the stumpy.match(target, ts) function, but apparently it is not suitable for multidimensional time series. Is there any way to solve this problem with stumpy?

Any help is appreciated!

**Comment by seanlaw** [comment, marked as answer]
@huberl Thank you for your question and welcome to the STUMPY community. Can you please tell me which version of STUMPY you currently have installed?

In Python:

```
import stumpy
print(stumpy.__version__)
```

I believe that `stumpy.match` should be able to handle multi-dimensional time series starting around version `1.11.x` (thanks to some excellent work by @mihailescum and @SaVoAMP) but would fail with older versions. However, please note that `stumpy.match` will compute the average distance (averaged across all 10 dimensions) as this is consistent with the definition put forth by the original "matrix profile" authors.

Alternatively, if this doesn't suit your needs then I recommend taking a look at the contents of the `stumpy.match` function and use it as a guide to customize as you see fit. It's not terribly intimidating and you can ask any clarifying questions here:

https://github.com/TDAmeritrade/stumpy/blob/4251ac398a53e8f4025a3ae2ebb3c6a53da2519a/stumpy/motifs.py#L419-L473

The most important thing is that `stumpy.match` just re-computes the 1-D matrix profiles one dimension at a time in:

https://github.com/TDAmeritrade/stumpy/blob/4251ac398a53e8f4025a3ae2ebb3c6a53da2519a/stumpy/motifs.py#L449-L453

And then it computes the mean across all dimensions. I hope this helps.

**Comment by huberl** [comment]
Thanks for the quick answer @seanlaw! It indeed was a version issue. With version 1.11.1 the `match` function can handle multidimensional time series.

As mentioned [here](https://stumpy.readthedocs.io/en/latest/Tutorial_Multidimensional_Motif_Discovery.html) and in the [Matrix Profile VI paper](https://www.cs.ucr.edu/~eamonn/Motif_Discovery_ICDM.pdf), simply averaging over all dimensions rarely produces good results. 

It probably makes sense to call it a match if at least k dimensions have a good match. What do you think about it?

**Reply by seanlaw** [reply]
> It probably makes sense to call it a match if at least k dimensions have a good match. What do you think about it?

It's really hard to know as it largely depends on the data. I recommend not overthinking it and just try it out and see what you find. In other words, test your hypothesis and iterate from there.

**Comment by huberl** [comment]
I modified the original code slightly, which gave pretty good results.

 ```
D = np.empty((d, n - m + 1))
for i in range(d):
    D[i, :] = core.mass(Q[i], T[i], M_T[i], Σ_T[i])

######## CUSTOM CODE ########
D = np.partition(D, k, axis=0)[:k]
#############################

D = np.mean(D, axis=0)
```
